### PR TITLE
[RHOL-329] Find cliques implementation in Scala

### PR DESCRIPTION
**Pull request by AbnerZheng**
## Overview
Equivalent to [networkx.find_cliques](https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.clique.find_cliques.html?highlight=find_clique#networkx.algorithms.clique.find_cliques) in Python

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/RHOL-329

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

**Comment by KentShikama** [comment]
WIP? Should I go ahead and give this a review?

**Comment by AbnerZheng** [comment]
@KentShikama  I think now it is ready to review.

File: casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala [hunk 1]
```diff
@@ -119,6 +119,24 @@ sealed abstract class SafetyOracleInstances {
...
+              if (weight > maxWeight) {
+                (clique, weight)
+              } else if (weight == maxWeight && clique.size > maxClique.size) {
+                (clique, weight)
+              } else {
+                (maxClique, maxWeight)
+              }
+            }
+          }
+          .pure[F]
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: There is no need to close this over F. Just have this as a pure function.

File: casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala [hunk 2]
```diff
@@ -225,7 +243,8 @@ sealed abstract class SafetyOracleInstances {
 
         for {
           edges <- computeAgreementGraphEdges
-        } yield edges.size
+          re    <- findMaximumClique(edges, candidates)
+        } yield re._1.size
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: And then here you can just do 
```
for {
  ...
} yield findMaximumClique(edges, candidates)._1.size

Comment (AbnerZheng) [reply casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: Done.

File: casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala (hunk 1 above)
Comment (AbnerZheng) [reply casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: Done

File: casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala [hunk 3]
```diff
@@ -119,6 +119,23 @@ sealed abstract class SafetyOracleInstances {
           }
         } yield mainParentWeightMap
 
+      private def findMaximumClique(edges: List[(Validator, Validator)],
+                            candidates: Map[Validator, Int]): (List[Validator], Int) =
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: Are you sure you ran the formatter?

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: sbt casper/test

Comment (AbnerZheng) [reply casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: Sorry about it. I have reformatted it.

**Pull request merged**